src/classes/RD2_ScheduleService.cls

```diff
@@ -511,13 +510,73 @@ public without sharing class RD2_ScheduleService {
         return nextDate;
     }
 
+    public List<ActiveSchedule> getVisualizedSchedules(npe03__Recurring_Donation__c rd) {
```

Comment:
Methods can be executed without additional checking since `getExistingSchedules()` already verifies for null `rd`. Also, if `schedules` are empty, the `for` loop will just exit since we are returning an empty list not a null now.
```
 public List<ActiveSchedule> getVisualizedSchedules(npe03__Recurring_Donation__c rd) {
        List<ActiveSchedule> activeSchedules = new List<ActiveSchedule>();
        List<RecurringDonationSchedule__c> schedules = getExistingSchedules(rd);
        
        Boolean first1stAnd15thFound = false;
        for (RecurringDonationSchedule__c schedule : schedules) {
```